src/mpol/crossval.py

```diff
@@ -0,0 +1,182 @@
+    diag_fig_train : bool, default=False
+        Whether to generate a diagnostic figure during training
+        (if True, `train_diag_step` must also be nonzero)
+    device : torch.device, default=None
+        Which hardware device to perform operations on (e.g., 'cuda:0').
+        'None' defaults to current device. 
+    verbose : bool, default=True
+        Whether to print notification messages. 
+    """
+    def __init__(self, coords, gridder, kfolds=5, seed=None, learn_rate=0.5, 
+                epochs=500, convergence_tol=1e-2, 
+                lambda_guess=None, lambda_guess_briggs=[0.0, 0.5], 
+                lambda_entropy=None, entropy_prior_intensity=1e-10, 
+                lambda_sparsity=None, lambda_TV=None, 
+                TV_epsilon=1e-10, lambda_TSV=None, 
+                train_diag_step=None, diag_fig_train=False, device=None, 
+                verbose=True):
+        self._coords = coords
+        self._gridder = gridder        
+        self._kfolds = kfolds
+        self._seed = seed
+        self._learn_rate = learn_rate
+        self._epochs = epochs
+        self._convergence_tol = convergence_tol
+        self._lambda_guess = lambda_guess
+        self._lambda_guess_briggs = lambda_guess_briggs
+        self._lambda_entropy = lambda_entropy
+        self._entropy_prior_intensity = entropy_prior_intensity
+        self._lambda_sparsity = lambda_sparsity
+        self._lambda_TV = lambda_TV
+        self._TV_epsilon = TV_epsilon
+        self._lambda_TSV = lambda_TSV
+        self._train_diag_step = train_diag_step
+        self._diag_fig_train = diag_fig_train
+        self._device = device
+        self._verbose = verbose
+
+
+    def split_dataset(self, dataset):
+        r"""
+        Split a dataset into training and test subsets. 
+
+        Parameters
+        ----------
+        dataset : PyTorch dataset object
+            Instance of the `mpol.datasets.GriddedDataset` class
+        kfolds : int 
+            Number of k-folds to use in cross-validation
+        seed : int 
+            Seed for random number generator used in splitting data
+
+        Returns
+        -------
+        test_train_datasets : list of `mpol.datasets.GriddedDataset` objects
+            Training and test subsets obtained from splitting the input dataset
+        """
+        # create a radial and azimuthal partition for the dataset
+        dartboard = Dartboard(coords=self._coords)
+
+        # use 'dartboard' to split full dataset into train/test subsets
+        subsets = KFoldCrossValidatorGridded(dataset, k=self._kfolds,
+                                        dartboard=dartboard,
+                                        npseed=self._seed)
+
+        if hasattr(self._device,'type') and self._device.type == 'cuda': # TODO: confirm which objects need to be passed to gpu
+            test_train_datasets = [(train.to(self._device), test.to(self._device)) for (train, test) in subsets]
```

Comment:
I think this will work OK for most things we're working on now, but I think the eventual model that we should adopt is to send one dataset to the GPU at a time, do the training for that, send the results back tothe CPU, and then repeat for the next dataset subset. Because `test_train_datasets` is now a list that of a bunch of K copies of the dataset, there's a good chance that it won't fit on GPU memory all at once especially if the original dataset is large.